File: CJHouser/python3/d04.py
import boiler
import re

def part1(fs, ps, part1):
    valid_count = 0
    for p in ps:
        valid = True
        i = 0
        while i < len(fs) and valid:
            if fs[i] not in p and fs[i] != "cid":
                valid = False
            i += 1
        valid_count += valid and (part1 or validate(p))
    return valid_count

def validate(p):
    v = True
    i = 0
    keys = list(p.keys())
    while i < len(keys) and v:
        if keys[i] != "cid":
            v = dispatch[keys[i]](p[keys[i]])
        i += 1
    return v

def checkHeight(n):
    v = True
    height = re.match("^[0-9]+[cm$|in$]", n)
    if not height:
        v = False
    else:
        if n[-2:] == "cm" and int(n[:-2]) not in range(150, 194):
            v = False
        if n[-2:] == "in" and int(n[:-2]) not in range(59, 77):
            v = False
    return v

dispatch = {
    "byr": (lambda n: int(n) in range(1920, 2003)),
    "iyr": (lambda n: int(n) in range(2010, 2021)),
    "eyr": (lambda n: int(n) in range(2020, 2031)),
    "hgt": checkHeight,
    "hcl": (lambda n: re.match("^#[0-9a-f]{6}$", n) != None),
    "ecl": (lambda n: n in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]),
    "pid": (lambda n: re.match("^[0-9]{9}$", n) != None),
}

if __name__ == "__main__":
    entries = boiler.fileContents("d04").split("\n\n")
    passports = []
    for e in entries:
        e_fields = re.findall("[a-zA-Z]+:#*[a-zA-Z0-9]*(?=[\s\n]*)", e)
        passport = dict()
        for ef in e_fields:
            key, value = ef.split(':')
            passport[key] = value
        passports.append(passport)
    fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid", "cid"]
    print(part1(fields, passports, True))
    print(part1(fields, passports, False))

File: CJHouser/haskell/d04.hs
import qualified Data.Text.IO as DTIO
import Text.Megaparsec
import Text.Megaparsec.Char
import Data.Text (Text)
import Data.Void
import Data.Char

-- Check if all required fields are present.
part1 :: [(String, String)] -> Bool
part1 xs = let f x = (&&) (elem x (map fst xs))
           in foldr f True ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

-- Check if required fields are valid.
part2 :: (String, String) -> Bool
part2 ("cid", val) = True
part2 ("byr", val) = length val == 4 && elem (read val :: Int) [1920..2002]
part2 ("iyr", val) = length val == 4 && elem (read val :: Int) [2010..2020]
part2 ("eyr", val) = length val == 4 && elem (read val :: Int) [2020..2030]
part2 ("hgt", val) = let value = read (takeWhile isDigit val) :: Int
                         unit  = dropWhile isDigit val
                     in (unit == "cm" && elem value [150..193])
                        || (unit == "in" && elem value [59..76])
part2 ("hcl", val) = and [head val == '#',
                          length val == 7,
                          let f x = (&&) (isHexDigit x && (isLower x || isDigit x))
                          in foldr f True (tail val)]
part2 ("ecl", val) = elem val ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
part2 ("pid", val) = length val == 9
                     && foldr (\x -> (&&) (isDigit x)) True val
otherwise          = False

type Parser = Parsec Void Text

-- Separate field name and value
fieldParser :: Parser (String, String)
fieldParser = (,) <$> someTill lowerChar (char ':') <*> some (char '#' <|> alphaNumChar)

-- Separates field/values within passport
passportParser :: Parser [(String, String)]
passportParser = endBy fieldParser (optional spaceChar)

-- Separates passports within input
inputParser :: Parser [[(String, String)]]
inputParser = sepBy passportParser spaceChar

main = do
    --parseTest fieldParser "aaa:#aaa000"
    --parseTest passportParser "aaa:#aaa000 bbb:#bbb111\nccc:#ccc222"
    --parseTest inputParser "aaa:#aaa000 bbb:#bbb111\nccc:#ccc222\n\nddd:#ddd333"
    let filename = "../inputs/d04"
    contents <- DTIO.readFile filename
    case runParser inputParser filename contents of
        Left e -> print e
        Right xss -> do let p1 = filter part1 xss
                        let p2 = filter (foldr (\x -> (&&) $ part2 x) True) p1
                        print $ length $ p1
                        print $ length $ p2
